### ISSUE TITLE
Multithreading issue: `Kratos` gets stuck when two instances runs at the same time

### ISSUE THREAD

**Issue by thelfer** [KratosMultiphysics/Kratos#6020]
`Kratos` gets stuck when two instances runs at the same time.

I am using Debian stretch with python 3.5.3. `strace` outputs show that this is related to multithreading:

```sh
futex(0x5590535e6294, FUTEX_WAIT_PRIVATE, 2110824, NULL) = -1 EAGAIN (Resource temporarily unavailable)
futex(0x5590535e6294, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x5590535e6294, FUTEX_WAKE_PRIVATE, 2147483647) = 2
futex(0x5590535e6294, FUTEX_WAKE_PRIVATE, 2147483647) = 3
futex(0x5590535e6294, FUTEX_WAKE_PRIVATE, 2147483647) = 0
```

The problem disapears when setting the `OMP_NUM_THREADS` before lauching each processes.

**Comment by mpentek** [comment]
DId you try running like:
   `taskset -c 0-3 python3 MainKratos.py` -> for one example
   `taskset -c 4-7 python3 MainKratos.py` -> for another example
when running 2 jobs concurrently.

Perhaps you can get help and in the long run there might be a correct way so that the software handles it internally, until then pinning OpenMP tasks to certain processors/threads might be a workaround.

And checking out OMP_NUM_THREADS setting.

**Comment by RiccardoRossi** [comment]
also, OMP_NUM_THREADS is normally set to the max number of avaialble processors, so maybe what you are experiencing is simply oversubscription

**Comment by philbucher** [comment]
I agree with @RiccardoRossi, this seems to be normal behavior
Looking at the error you get ("(Resource temporarily unavailable)") I would say this makes sense

**Comment by thelfer** [comment]
@mpentek Indeed, wtih your proposal it works (sorry, I missed your post).
@philbucher, @RiccardoRossi No doubt that you're technically right, but this is awkward from the user point of view.
@pooyan-dadvand Shall I close this issue ?

**Comment by RiccardoRossi** [comment]
@thelfer i am not sure of what a fix would be in this case. what do you propose?

the openmp way of controlling the number of threads is to set the env variable...

**Comment by thelfer** [comment]
It is clear that I am facing oversubscription. However, as a user, I would have expected overall system slowdown, not a freeze of `Kratos` processes. I ask a colleague of mine which has much more experience with `OpenMP` than I do. This is also what he would have expected, but that is behaviour sadly frequently arises.
 
More precisely, the trouble arises in the `amgcl` library which has short parallel sections ( "#pragma omp parallel")  closed by synchronisation barriers "#pragam omp barrier". See  https://github.com/ddemidov/amgcl/blob/master/amgcl/relaxation/detail/ilu_solve.hpp

In conclusion, this may be more related to this library than to `Kratos`.

Thanks for the discussion.

**Comment by loumalouomega** [comment]
> 
> 
> It is clear that I am facing oversubscription. However, as a user, I would have expected overall system slowdown, not a freeze of `Kratos` processes. I ask a colleague of mine which has much more experience with `OpenMP` than I do. This is also what he would have expected, but that is behaviour sadly frequently arises.
> 
> More precisely, the trouble arises in the `amgcl` library which has short parallel sections ( "#pragma omp parallel") closed by synchronisation barriers "#pragam omp barrier". See https://github.com/ddemidov/amgcl/blob/master/amgcl/relaxation/detail/ilu_solve.hpp
> 
> In conclusion, this may be more related to this library than to `Kratos`.
> 
> Thanks for the discussion.

FYI @ddemidov

**Comment by ddemidov** [comment]
Could this discussion be relevant?
https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/733576

Especially this bit:

> You need to increase:
> - A stack size for OpenMP threads ( a default value is exceeded in your case ), and
> - A default stack size of your application ( Note: On a KNL server with a Linux Red Hat that I use 'ulimit -s' shows 512K ).

Also, it seems there is a problem with your implementation of OpenMP: it looks like one of the threads in `omp parallel` failed to be created due to some resource limitation, and the next `omp barrier` hangs (because the failed thread did not arrive?). Looks like the correct behavior for the program here would be to crash before entering the barrier. I don't think there is anything to do for me on amgcl side of things here.

**Comment by thelfer** [comment]
@ddemidov I increased both the stack sizes of the applications and the OpenMP threads without success.

Note that if I kill one of the two processes (or attach one in `gdb`), the other (re)-starts working as expected. So I don't expect any stack exhausion here.

**Comment by ddemidov** [comment]
Are those MPI processes, or just independent processes? In the latter case, I am not sure how can the behavior of one process can affect the other.

**Comment by thelfer** [comment]
@ddemidov Independent processes. Hence the issue :)

**Comment by ddemidov** [comment]
I think this issue should go to the OpenMP implementation team then.

**Comment by thelfer** [comment]
@ddemidov Probably. Sadly, I won't have the time to make a unit test...
